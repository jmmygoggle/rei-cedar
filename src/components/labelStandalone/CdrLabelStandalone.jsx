import style from './styles/CdrLabelStandalone.scss';

export default {
  name: 'CdrLabelStandalone',
  props: {
    forId: {
      type: String,
      required: true,
    },
    label: String,
    disabled: Boolean,
    required: Boolean,
    optional: Boolean,
    hideLabel: Boolean,
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    labelClass() {
      return {
        [this.style['cdr-label-standalone__label']]: true,
        [this.style['cdr-label-standalone__label--disabled']]: this.disabled,
        [this.style['cdr-label-standalone__label--sr-only']]: this.hideLabel,
      };
    },
    inputClass() {
      const hasLabelContent = !this.hideLabel || this.$slots.helper || this.$slots.info;
      return {
        [this.style['cdr-label-standalone__input-wrap']]: true,
        [this.style['cdr-label-standalone__input-spacing']]: hasLabelContent,
      };
    },
    labelEl() {
      const requiredEl = this.required ? (
        <span aria-label="required">
          *
        </span>
      ) : '';

      const optionalEl = this.optional ? (
        <span
          class={this.style['cdr-label-standalone__optional']}
        >
          (optional)
        </span>
      ) : '';

      return (
        <label
          class={this.labelClass}
          for={this.forId}
          ref="label"
        >
          { this.label }{ requiredEl || optionalEl ? ' ' : ''}{ requiredEl || optionalEl }
        </label>
      );
    },
  },
  render() {
    return (
      <div class={this.style['cdr-label-standalone']}>
        <div class={this.style['cdr-label-standalone__label-wrapper']}>
          { this.labelEl }
          { !this.hideLabel && this.$slots.helper && (<br/>) }
          { this.$slots.helper && (
            <span
              class={this.style['cdr-label-standalone__helper']}
              id={`${this.forId}-helper-text-top`}
            >
              { this.$slots.helper }
            </span>
          )}
        </div>

        <div class={this.inputClass}>
          {this.$slots.default}
          {this.$slots['info-action'] && (
            <div
              class={this.style['cdr-label-standalone__info-action']}
            >
              {this.$slots['info-action']}
            </div>
          )}
        </div>

        {this.$slots.info && (
          <span
            class={this.style['cdr-label-standalone__info']}
          >
            { this.$slots.info }
          </span>
        )}

        <div class={this.style['cdr-label-standalone__post-content']}>
          {this.$slots['helper-text-bottom']}
          {this.$slots.error}
        </div>
      </div>
    );
  },
};
