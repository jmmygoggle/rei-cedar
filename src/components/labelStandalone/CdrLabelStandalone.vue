<template>
  <div :class="[$style[baseClass], $style[wrapperClass]]">
    <label
      v-if="!hideLabel"
      :class="[$style['cdr-label-standalone__label'], $style[disabledLabelClass]]"
      :for="forId"
    >
      {{ label }}{{ required || optional ? ' ' : '' }}
      <span
        v-if="required"
        aria-label="required"
      >
        *
      </span>

      <span
        v-else-if="optional"
        :class="$style['cdr-label-standalone__optional']"
      >
        (optional)
      </span>
    </label>

    <br v-if="!hideLabel && hasHelper">
    <span
      :class="$style['cdr-label-standalone__helper']"
      v-if="hasHelper"
    >
      <slot name="helper" />
    </span>

    <span
      v-if="hasInfo"
      :class="$style['cdr-label-standalone__info']"
    >
      <slot name="info" />
    </span>
  </div>
</template>
<script lang="ts">
import { defineComponent, computed } from 'vue';

export default defineComponent({
  name: 'CdrLabelStandalone',
  props: {
    forId: {
      type: String,
      required: true,
    },
    label: String,
    disabled: Boolean,
    required: Boolean,
    optional: Boolean,
    hideLabel: Boolean,
  },
  setup(props, ctx) {
    const baseClass = 'cdr-label-standalone';
    const hasHelper = ctx.slots.helper;
    const hasInfo = ctx.slots.info;
    const disabledLabelClass = computed(() => props.disabled
      && 'cdr-label-standalone__label--disabled');
    const wrapperClass = computed(() => (!props.hideLabel || hasHelper || hasInfo)
      && 'cdr-label-standalone--spacing');
    return {
      hasHelper,
      hasInfo,
      baseClass,
      disabledLabelClass,
      wrapperClass,
    };
  },
});

</script>

<style lang="scss" module src="./styles/CdrLabelStandalone.scss">
</style>
