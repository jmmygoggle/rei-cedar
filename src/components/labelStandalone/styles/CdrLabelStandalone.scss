@import '../../../css/settings/index.scss';
@import './vars/CdrLabelStandalone.vars.scss';

.cdr-label-standalone {
  position: relative;

  &--spacing {
    margin-bottom: $cdr-space-half-x;
  }

  &__label {
    @include cdr-label-standalone-label-mixin;
    display: inline-block;

    &--disabled {
      @include cdr-label-standalone-disabled-mixin;

      .cdr-label-standalone__optional {
        @include cdr-label-standalone-disabled-mixin;
      }
    }
  }

  &__optional {
    @include cdr-label-standalone-optional-mixin;
  }

  &__helper {
    @include cdr-label-standalone-helper-mixin;
    display: inline-block;
  }

  &__info {
    @include cdr-label-standalone-info-mixin;
    position: absolute;
    right: 0;
    bottom: 0;
  }

  &__info-action {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 40px;
  }

  &__input-wrap {
    position: relative;
    display: flex;
  }
}
