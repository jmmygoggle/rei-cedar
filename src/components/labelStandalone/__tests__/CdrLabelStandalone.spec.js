import { shallowMount } from '../../../../test/vue-jest-style-workaround.js';
import CdrLabelStandalone from 'componentdir/labelStandalone/CdrLabelStandalone';

describe('CdrFormLabelStandalone', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'Label Test',
        forId: 'test',
        required: true,
      },
      slots: {
        helper: 'very helpful',
        info: 'very informative',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a label element', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'Label Test',
        forId: 'test',
      },
    });
    expect(wrapper.find('label').textContent).toBe('Label Test');
  });

  it('does not render label if hideLabel is passed', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'Label Test',
        forId: 'test',
        hideLabel: true,
      },
    });
    expect(wrapper.find('label').exists()).toBe(false);
  });

  it('maps input id to label for correctly', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'testing',
        forId: 'test',
      },
    });
    expect(wrapper.find('label').htmlFor).toBe('test');
  });

  it('renders required label', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'test',
        required: true,
        forId: 'test',
      },
    });
    expect(wrapper.find('label').textContent).toBe('test *');
  });

  it('does not render both required and optional labels simultaneously', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'test',
        required: true,
        optional: true,
        forId: 'test',
      },
    });
    expect(wrapper.find('label').textContent).toBe('test *');
  });

  it('renders optional messaging', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'test',
        optional: true,
        forId: 'test',
      },
    });
    expect(wrapper.find('label').textContent).toBe('test (optional)');
  });

  it('renders helper slot', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'test',
        forId: 'test',
      },
      slots: {
        'helper': 'very helpful',
      },
    });
    expect(wrapper.find('.cdr-label-standalone__helper').text()).toBe('very helpful');
  });

  it('renders break between label and helper if both are present', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'test',
        forId: 'test',
      },
      slots: {
        'helper': 'very helpful',
      },
    });
    expect(wrapper.find('br').exists()).toBe(true);
  });

  it('does not render break between label and helper if label is hidden', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'test',
        hideLabel: true,
        forId: 'test',
      },
      slots: {
        'helper': 'very helpful',
      },
    });
    expect(wrapper.find('br').exists()).toBe(false);
  });

  it('renders info slot', () => {
    const wrapper = shallowMount(CdrLabelStandalone, {
      propsData: {
        label: 'test',
        forId: 'test',
      },
      slots: {
        info: 'very informative',
      },
    });
    expect(wrapper.find('.cdr-label-standalone__info').text()).toBe('very informative');
  });
});
