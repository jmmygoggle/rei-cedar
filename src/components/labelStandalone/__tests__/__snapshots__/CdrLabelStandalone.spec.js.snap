// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrFormLabelStandalone matches snapshot 1`] = `
<div>
  <div
    class="cdr-label-standalone cdr-label-standalone--spacing"
  >
    <label
      class="cdr-label-standalone__label"
      for="test"
    >
      Label Test 
      <span
        aria-label="required"
      >
        *
      </span>
    </label>
    <br />
    <span
      class="cdr-label-standalone__helper"
    >
      very helpful
    </span>
    <span
      class="cdr-label-standalone__info"
    >
      very informative
    </span>
  </div>
  <div
    class="cdr-label-standalone__input-wrap"
  />
</div>
`;
